pkg/integrations/register.go

```diff
@@ -0,0 +1,240 @@
+package integrations
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	registeredIntegrations = []Config{}
+	configFieldNames       = make(map[reflect.Type]string)
+
+	emptyStructType = reflect.TypeOf(struct{}{})
+	configsType     = reflect.TypeOf(Configs{})
+)
+
+// RegisterIntegration dynamically registers a new integration. The Config
+// will represent the configuration that controls the specific integration.
+// Registered Configs may be loaded using UnmarshalYAML or manually
+// constructed.
+//
+// RegisterIntegration panics if cfg is not a pointer.
+func RegisterIntegration(cfg Config) {
+	if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
+		panic(fmt.Sprintf("RegisterIntegration must be given a pointer, got %T", cfg))
+	}
+	registeredIntegrations = append(registeredIntegrations, cfg)
+	configFieldNames[reflect.TypeOf(cfg)] = cfg.Name()
+}
+
+// Configs is a list of integrations.
+type Configs []Config
+
+func (c *Configs) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return c.unmarshalWithIntegrations(registeredIntegrations, unmarshal)
+}
+
+func (c *Configs) unmarshalWithIntegrations(integrations []Config, unmarshal func(interface{}) error) error {
+	// Create a dynamic struct type full of our registered integrations and
+	// unmarshal to it.
+	var fields []reflect.StructField
+	for _, cfg := range integrations {
+		fields = append(fields, reflect.StructField{
+			Name: "Config_" + cfg.Name(),
+			Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s"`, cfg.Name())),
+			Type: reflect.TypeOf(cfg),
+		})
+	}
+
+	var (
+		structType = reflect.StructOf(fields)
+		structVal  = reflect.New(structType)
+	)
+	if err := unmarshal(structVal.Interface()); err != nil {
+		return err
+	}
+
+	// Go over all non-nil fields in structVal and append them to c.
+	structVal = structVal.Elem()
+	for i := 0; i < structVal.NumField(); i++ {
+		if structVal.Field(i).IsNil() {
+			continue
+		}
+
+		val := structVal.Field(i).Interface().(Config)
+		*c = append(*c, val)
+	}
+
+	return nil
+}
+
+// MarshalYAML helps implement yaml.Marshaller for structs that have a Configs
+// field that should be inlined in the YAML string.
+func MarshalYAML(v interface{}) (interface{}, error) {
+	inVal := reflect.ValueOf(v)
+	for inVal.Kind() == reflect.Ptr {
+		inVal = inVal.Elem()
+	}
+	if inVal.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("integrations: can only marshal a struct, got %T", v)
+	}
+	inType := inVal.Type()
+
+	var (
+		cfgType    = getConfigTypeForIntegrations(registeredIntegrations, inType)
+		cfgPointer = reflect.New(cfgType)
+		cfgVal     = cfgPointer.Elem()
+	)
+
+	// Copy over any existing value from inVal to cfgVal.
+	//
+	// The ordering of fields in inVal and cfgVal match identically up until the
+	// extra fields appended to the end of cfgVal.
+	var configs Configs
+	for i, n := 0, inType.NumField(); i < n; i++ {
+		if inType.Field(i).Type == configsType {
+			configs = inVal.Field(i).Interface().(Configs)
```

Comment:
could you explain: this looks like we could be overwriting configs, unless we are only expecting one field of this type? i was expecting to see an `append` in here

Reply:
Ah from line 106 it looks like we're only expecting one of these.